### PR TITLE
[expo-apple-auth] Add “Sign up with Apple” option

### LINKED ISSUE #7454: AppleAuthentication – where is the “Sign up with Apple” option?
Hi, everyone.

I’m looking at using the `AppleAuthentication` module and I noticed that the only options for the text of the button are “Sign **in** with Apple” and “Continue with Apple”. There’s no option for “Sign **up** with Apple”, which is another option shown [here](https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/buttons/#creating-a-custom-sign-in-with-apple-button) in Apple’s Human Interface Guidelines. Here's what Apple show in their documentation.

Screenshot 2020-03-24 at 09 54 51

The `AppleAuthentication.AppleAuthenticationButtonType` enum type [only offers](https://github.com/expo/expo/blob/sdk-36/packages/expo-apple-authentication/src/AppleAuthentication.types.ts#L213-L216) the two I mentioned. Is there a reason the sign **up** version isn’t available?

Thanks!

### PR DESCRIPTION
**Pull request by IjzerenHein**
# Why

Add support for "Sign **up** with Apple" to `expo-apple-authentication`.
This option was added by Apple in iOS 13.2, but wasn't available to our users yet.

# How

Added this option to the `expo-apple-authentication` package. As well as the NCL and to the docs.

Renders an empty view when used on iOS 13.1 or lower.

# Test Plan

New option has been added to the NCL test app:

image
image
image

Because this option is only available on iOS 13.2, running this on a lower iOS version will render an empty view (screenshot is from iOS 13.1):

image

# Other

Fixes #7454

> PR was created together with @jkhales during a pair programming session.

### CODE REVIEW

**Review by brentvatne — APPROVED** [review]

**Comment by brentvatne** [comment]
we should wait until after sdk 37 is released before merging this because we don't want to pull the docs changes in from unversioned to 37

**Comment by IjzerenHein** [comment]
The iOS build is currently failing because the new `SignUp` constants are not recognized on CI.
It gives this error:

```
/Users/distiller/project/packages/expo-apple-authentication/ios/EXAppleAuthentication/EXAppleAuthenticationButtonViewManagers.m
:
83
:
1
:
 
error
: 
use of undeclared identifier 'ASAuthorizationAppleIDButtonTypeSignUp'; did you mean 'ASAuthorizationAppleIDButtonTypeSignIn'?

EX_REGISTER_APPLE_AUTHENTICATION_VIEW_MANAGER(SignUp, White, ExpoAppleAuthenticationButtonSignUpWhite, 13.2)

^

/Users/distiller/project/packages/expo-apple-authentication/ios/EXAppleAuthentication/EXAppleAuthenticationButtonViewManagers.m
:
22
:
136
:
 
note
: 
expanded from macro 'EX_REGISTER_APPLE_AUTHENTICATION_VIEW_MANAGER'

  - (UIView *)view { if (@available(iOS ios_version, *)) { return [[EXAppleAuthenticationButton alloc] initWithAuthorizationButtonType:ASAuthorizationAppleIDButtonType ## type authorizationButtonStyle:ASAuthorizationAppleIDButtonStyle ## style]; } else { return nil; } } \

                                                                                                                                       ^

<scratch space>
:
79
:
1
:
 
note
: 
expanded from here

ASAuthorizationAppleIDButtonTypeSignUp

^

In module '
AuthenticationServices
' imported from 
/Users/distiller/project/packages/expo-apple-authentication/ios/EXAppleAuthentication/EXAppleAuthenticationButton.h
:
5
:
/Applications/Xcode-11.GM.Seed.2.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.0.sdk/System/Library/Frameworks/AuthenticationServices.framework/Headers/ASAuthorizationAppleIDButton.h
```

I think we need to upgrade XCode on the build server, What do you think @sjchmiela ?

**Comment by esamelson** [comment]
@nicknovitski would you be able to help out with upgrading Xcode in CI, when you get a chance? for this PR we need at least Xcode 11.2.

**Comment by tsapeta** [comment]
@IjzerenHein I've upgraded Xcode on the CI to 11.4.0 (#7316), it's not merged yet but should be soon 😉

File: packages/expo-apple-authentication/src/AppleAuthentication.types.ts [hunk 1]
```diff
@@ -213,6 +213,7 @@ export enum AppleAuthenticationUserDetectionStatus {
 export enum AppleAuthenticationButtonType {
   SIGN_IN = 0,
   CONTINUE = 1,
+  SIGN_UP = 2, // requires iOS 13.2 or later
```
Comment (tsapeta) [inline packages/expo-apple-authentication/src/AppleAuthentication.types.ts:216]: Might be better to put this information as tsdoc comment, so people are more likely to see this when they use it

**Review by tsapeta — APPROVED** [review]
Approving, but please add changelog (`packages/expo-apple-authentication/CHANGELOG.md`) entry as well 🙏

**Comment by darrylyoung** [comment]
Thanks for making this happen, guys.

It wasn't that long ago when I reported this so I really appreciate the quick turnaround. 🙏

File: packages/expo-apple-authentication/src/AppleAuthentication.types.ts (hunk 1 above)
Comment (IjzerenHein) [reply packages/expo-apple-authentication/src/AppleAuthentication.types.ts]: Good suggestion, thanks!

**Comment by darrylyoung** [comment]
Hi, @IjzerenHein. Should this be part of the recently-released `2.1.1`? I just updated and I see that on `master` the new option type is [there](https://github.com/expo/expo/blob/master/packages/expo-apple-authentication/src/AppleAuthentication.types.ts#L213-L220) but I'm not seeing it locally. If I dig into the `node_modules`, I can see that `AppleAuthentication.types.d.ts` still looks like this:

```ts
export declare enum AppleAuthenticationButtonType {
    SIGN_IN = 0,
    CONTINUE = 1
}
```

Am I missing something? Thank you.

**Comment by IjzerenHein** [comment]
Hi Darryl. No this has not been released yet.

**Comment by darrylyoung** [comment]
> Hi Darryl. No this has not been released yet.

Got it. Thanks anyway. Have a good week. ✌️

**Comment by darrylyoung** [comment]
Hi again, @IjzerenHein. Will this be released at some point soon? Thanks!

**Comment by IjzerenHein** [comment]
Yep this will be released in SDK 38 which is just about to enter QA and testing. Not sure what the exact release date will be, but should be within a couple weeks max

**Comment by darrylyoung** [comment]
Great. Thank you, @IjzerenHein. Much appreciated.

**Comment by raRaRa** [comment]
Hey all. Is there a good way to test the configuration for Sign in with Apple through the Expo app on iOS? I tried calling the https://appleid.apple.com/auth/token endpoint to get access & refresh token, but I keep getting `invalid_client` error. What would be the best way to test my config? Should I use TestFlight?

Thanks!

**Pull request merged**